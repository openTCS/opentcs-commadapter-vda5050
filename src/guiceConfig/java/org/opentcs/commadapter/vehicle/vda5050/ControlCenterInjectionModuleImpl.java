// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.commadapter.vehicle.vda5050;

import org.opentcs.customizations.controlcenter.ControlCenterInjectionModule;

/**
 * A custom Guice module for project-specific configuration.
 */
public class ControlCenterInjectionModuleImpl
    extends
      ControlCenterInjectionModule {

  /**
   * Creates a new instance.
   */
  public ControlCenterInjectionModuleImpl() {
  }

  @Override
  protected void configure() {
    install(new org.opentcs.commadapter.vehicle.vda5050.v1_1.ControlCenterInjectionModuleImpl());
    install(new org.opentcs.commadapter.vehicle.vda5050.v2_0.ControlCenterInjectionModuleImpl());
  }
}
