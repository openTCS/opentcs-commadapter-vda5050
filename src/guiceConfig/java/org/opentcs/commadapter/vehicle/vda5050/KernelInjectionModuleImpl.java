/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.commadapter.vehicle.vda5050;

import javax.inject.Singleton;
import org.opentcs.commadapter.vehicle.vda5050.common.mqtt.MqttClientManager;
import org.opentcs.commadapter.vehicle.vda5050.common.mqtt.MqttConfiguration;
import org.opentcs.customizations.kernel.KernelInjectionModule;

public class KernelInjectionModuleImpl
    extends
      KernelInjectionModule {

  /**
   * Creates a new instance.
   */
  public KernelInjectionModuleImpl() {
  }

  @Override
  protected void configure() {
    configureMqtt();

    bind(CommAdapterConfiguration.class)
        .toInstance(
            getConfigBindingProvider().get(
                CommAdapterConfiguration.PREFIX,
                CommAdapterConfiguration.class
            )
        );

    install(new org.opentcs.commadapter.vehicle.vda5050.v1_1.KernelInjectionModuleImpl());
    install(new org.opentcs.commadapter.vehicle.vda5050.v2_0.KernelInjectionModuleImpl());

    vehicleCommAdaptersBinder().addBinding().to(CommAdapterFactoryImpl.class);
  }

  private void configureMqtt() {
    bind(MqttConfiguration.class).toInstance(
        getConfigBindingProvider().get(MqttConfiguration.PREFIX, MqttConfiguration.class)
    );
    bind(MqttClientManager.class).in(Singleton.class);
  }
}
