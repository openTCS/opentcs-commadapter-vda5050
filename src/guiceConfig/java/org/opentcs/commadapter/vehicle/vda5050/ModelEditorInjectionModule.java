// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.commadapter.vehicle.vda5050;

import org.opentcs.customizations.plantoverview.PlantOverviewInjectionModule;

public class ModelEditorInjectionModule
    extends
      PlantOverviewInjectionModule {

  /**
   * Creates a new instance.
   */
  public ModelEditorInjectionModule() {
  }

  @Override
  protected void configure() {
    install(new org.opentcs.commadapter.vehicle.vda5050.v1_1.ModelEditorInjectionModule());
    install(new org.opentcs.commadapter.vehicle.vda5050.v2_0.ModelEditorInjectionModule());
  }
}
