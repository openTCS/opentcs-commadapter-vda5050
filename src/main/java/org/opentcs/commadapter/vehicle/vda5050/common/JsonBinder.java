/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.commadapter.vehicle.vda5050.common;

import static java.util.Objects.requireNonNull;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import java.io.IOException;
import java.util.function.Function;
import javax.annotation.Nonnull;

/**
 * Binds JSON strings to objects and vice versa.
 * Optionally, a filter can be applied before serializing to JSON.
 */
public class JsonBinder {

  /**
   * Maps between objects and their JSON representations.
   */
  private final ObjectMapper objectMapper
      = new ObjectMapper()
          .registerModule(new JavaTimeModule())
          .disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);

  private Function<JsonNode, JsonNode> filter = Function.identity();

  /**
   * Creates a new instance.
   */
  public JsonBinder() {
  }

  /**
   * Sets the filter that will be applied before serializing to JSON.
   *
   * @param filter The filter to be applied.
   */
  public void setFilter(
      @Nonnull
      Function<JsonNode, JsonNode> filter
  ) {
    this.filter = requireNonNull(filter, "filter");
  }

  /**
   * Maps the given JSON string to an object.
   *
   * @param <T> The type of object to map to.
   * @param jsonString The JSON string.
   * @param clazz The type of object to map to.
   * @return The object created from the JSON string.
   * @throws IllegalArgumentException In case there was a problem mapping the given object from
   * JSON.
   */
  public <T> T fromJson(String jsonString, Class<T> clazz)
      throws IllegalArgumentException {
    try {
      return objectMapper.readValue(jsonString, clazz);
    }
    catch (IOException exc) {
      throw new IllegalArgumentException("Could not parse JSON input", exc);
    }
  }

  /**
   * Applies the filter set via {@link #setFilter(Function)} and maps the given
   * object to a JSON string.
   *
   * @param object The object to be mapped.
   * @return The JSON string representation of the object.
   * @throws IllegalArgumentException In case there was a problem mapping the given object to JSON.
   */
  public String toJson(Object object)
      throws IllegalArgumentException {
    try {
      return objectMapper
          .writerWithDefaultPrettyPrinter()
          .writeValueAsString(filter.apply(objectMapper.valueToTree(object)));
    }
    catch (JsonProcessingException exc) {
      throw new IllegalArgumentException("Could not produce JSON output", exc);
    }
  }
}
