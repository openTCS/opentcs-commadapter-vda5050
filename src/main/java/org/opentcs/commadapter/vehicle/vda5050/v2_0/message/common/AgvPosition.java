/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.commadapter.vehicle.vda5050.v2_0.message.common;

import static java.util.Objects.requireNonNull;
import static org.opentcs.commadapter.vehicle.vda5050.common.Assertions.checkInRange;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import java.io.Serializable;
import javax.annotation.Nonnull;

/**
 * Defines the position on a map in world coordinates. Each floor has its own map.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
public class AgvPosition
    implements
      Serializable {

  /**
   * X-position on the map in reference to the map coordinate system.
   * <p>
   * Precision is up to the specific implementation.
   */
  private Double x;
  /**
   * Y-position on the map in reference to the map coordinate system.
   * <p>
   * Precision is up to the specific implementation.
   */
  private Double y;
  /**
   * Orientation of the AGV (in rad). Range: [-PI ... PI]
   */
  private Double theta;
  /**
   * Unique identification of the map in which the position is referenced.
   * <p>
   * Each map has the same origin of coordinates. When an AGV uses an elevator, e. g. leading from a
   * departure floor to a target floor, it will disappear off the map of the departure floor and
   * spawn in the related lift node on the map of the target floor.
   */
  private String mapId;
  /**
   * [Optional] Additional information on the map.
   */
  private String mapDescription;
  /**
   * Whether the AGV's position is initialized or not.
   */
  private Boolean positionInitialized;
  /**
   * [Optional] Describes the quality of the localization. Range: [0.0 ... 1.0]
   * <p>
   * Can be used e. g. by SLAM-AGVs to describe how accurate the current position information is.
   * <p>
   * Interpretation of values:
   * <ul>
   * <li>0.0: Position unknown</li>
   * <li>1.0: Position known</li>
   * </ul>
   * <p>
   * Optional for vehicles that cannot estimate their localization score. Only for logging and
   * visualization purposes.
   */
  private Double localizationScore;
  /**
   * [Optional] Value for position deviation range (in m).
   * <p>
   * Optional for vehicles that cannot estimate their deviation (e.g. with grid-based localization).
   * Only for logging and visualization purposes.
   */
  private Double deviationRange;

  // CHECKSTYLE:OFF (Long lines because some of these parameter declarations are very long.)
  @JsonCreator
  public AgvPosition(
      @Nonnull
      @JsonProperty(required = true, value = "x")
      Double x,
      @Nonnull
      @JsonProperty(required = true, value = "y")
      Double y,
      @Nonnull
      @JsonProperty(required = true, value = "theta")
      Double theta,
      @Nonnull
      @JsonProperty(required = true, value = "mapId")
      String mapId,
      @Nonnull
      @JsonProperty(required = true, value = "positionInitialized")
      Boolean positionInitialized
  ) {
    this.x = requireNonNull(x, "x");
    this.y = requireNonNull(y, "y");
    this.theta = checkInRange(requireNonNull(theta, "theta"), -Math.PI, Math.PI, "theta");
    this.mapId = requireNonNull(mapId, "mapId");
    this.positionInitialized = requireNonNull(positionInitialized, "positionInitialized");
  }
  // CHECKSTYLE:ON

  public Double getX() {
    return x;
  }

  public AgvPosition setX(
      @Nonnull
      Double x
  ) {
    this.x = requireNonNull(x, "x");
    return this;
  }

  public Double getY() {
    return y;
  }

  public AgvPosition setY(
      @Nonnull
      Double y
  ) {
    this.y = requireNonNull(y, "y");
    return this;
  }

  public Double getTheta() {
    return theta;
  }

  public AgvPosition setTheta(
      @Nonnull
      Double theta
  ) {
    this.theta = checkInRange(requireNonNull(theta, "theta"), -Math.PI, Math.PI, "theta");
    return this;
  }

  public String getMapId() {
    return mapId;
  }

  public AgvPosition setMapId(
      @Nonnull
      String mapId
  ) {
    this.mapId = requireNonNull(mapId, "mapId");
    return this;
  }

  public String getMapDescription() {
    return mapDescription;
  }

  public AgvPosition setMapDescription(String mapDescription) {
    this.mapDescription = mapDescription;
    return this;
  }

  public Boolean isPositionInitialized() {
    return positionInitialized;
  }

  public AgvPosition setPositionInitialized(
      @Nonnull
      Boolean positionInitialized
  ) {
    this.positionInitialized = requireNonNull(positionInitialized, "positionInitialized");
    return this;
  }

  public Double getLocalizationScore() {
    return localizationScore;
  }

  public AgvPosition setLocalizationScore(Double localizationScore) {
    if (localizationScore != null) {
      checkInRange(localizationScore, 0.0, 1.0, "localizationScore");
    }
    this.localizationScore = localizationScore;
    return this;
  }

  public Double getDeviationRange() {
    return deviationRange;
  }

  public AgvPosition setDeviationRange(Double deviationRange) {
    this.deviationRange = deviationRange;
    return this;
  }

  @Override
  public String toString() {
    return "AgvPosition{" + "x=" + x
        + ", y=" + y
        + ", theta=" + theta
        + ", mapId=" + mapId
        + ", mapDescription=" + mapDescription
        + ", positionInitialized=" + positionInitialized
        + ", localizationScore=" + localizationScore
        + ", deviationRange=" + deviationRange
        + '}';
  }

}
