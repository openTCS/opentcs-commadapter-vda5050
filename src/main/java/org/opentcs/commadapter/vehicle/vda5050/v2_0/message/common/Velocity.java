/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.commadapter.vehicle.vda5050.v2_0.message.common;

import com.fasterxml.jackson.annotation.JsonInclude;
import java.io.Serializable;

/**
 * The AGV's velocity in vehicle coordinates.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
public class Velocity
    implements
      Serializable {

  /**
   * [Optional] The AGV's velocity in its x direction.
   */
  private Double vx;
  /**
   * [Optional] The AGV's velocity in its y direction.
   */
  private Double vy;
  /**
   * [Optional] The AGV's turning speed around its z axis.
   */
  private Double omega;

  public Velocity() {
  }

  public Double getVx() {
    return vx;
  }

  public Velocity setVx(Double vx) {
    this.vx = vx;
    return this;
  }

  public Double getVy() {
    return vy;
  }

  public Velocity setVy(Double vy) {
    this.vy = vy;
    return this;
  }

  public Double getOmega() {
    return omega;
  }

  public Velocity setOmega(Double omega) {
    this.omega = omega;
    return this;
  }

  @Override
  public String toString() {
    return "Velocity{" + "vx=" + vx + ", vy=" + vy + ", omega=" + omega + '}';
  }

}
