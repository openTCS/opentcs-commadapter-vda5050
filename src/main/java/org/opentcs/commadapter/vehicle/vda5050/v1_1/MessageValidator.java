/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.commadapter.vehicle.vda5050.v1_1;

import java.io.InputStreamReader;
import java.util.Map;
import org.opentcs.commadapter.vehicle.vda5050.common.JsonValidator;
import org.opentcs.commadapter.vehicle.vda5050.v1_1.message.connection.Connection;
import org.opentcs.commadapter.vehicle.vda5050.v1_1.message.instantactions.InstantActions;
import org.opentcs.commadapter.vehicle.vda5050.v1_1.message.order.Order;
import org.opentcs.commadapter.vehicle.vda5050.v1_1.message.state.State;
import org.opentcs.commadapter.vehicle.vda5050.v1_1.message.visualization.Visualization;

/**
 * A {@link JsonValidator} for VDA5050 1.1 messages.
 */
public class MessageValidator
    extends
      JsonValidator {

  /**
   * Creates a new instance.
   *
   * @throws IllegalArgumentException If there was a problem reading the VDA5050 schemas.
   */
  public MessageValidator()
      throws IllegalArgumentException {
    super(
        Map.of(
            Connection.class,
            new InputStreamReader(
                Connection.class.getResourceAsStream(Connection.JSON_SCHEMA_PATH)
            ),
            InstantActions.class,
            new InputStreamReader(
                InstantActions.class.getResourceAsStream(InstantActions.JSON_SCHEMA_PATH)
            ),
            Order.class,
            new InputStreamReader(
                Order.class.getResourceAsStream(Order.JSON_SCHEMA_PATH)
            ),
            State.class,
            new InputStreamReader(
                State.class.getResourceAsStream(State.JSON_SCHEMA_PATH)
            ),
            Visualization.class,
            new InputStreamReader(
                Visualization.class.getResourceAsStream(Visualization.JSON_SCHEMA_PATH)
            )
        )
    );
  }

}
