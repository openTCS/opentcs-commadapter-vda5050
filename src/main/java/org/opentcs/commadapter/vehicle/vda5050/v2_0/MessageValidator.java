// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.commadapter.vehicle.vda5050.v2_0;

import java.io.InputStreamReader;
import java.util.Map;
import org.opentcs.commadapter.vehicle.vda5050.common.JsonValidator;
import org.opentcs.commadapter.vehicle.vda5050.v2_0.message.connection.Connection;
import org.opentcs.commadapter.vehicle.vda5050.v2_0.message.instantactions.InstantActions;
import org.opentcs.commadapter.vehicle.vda5050.v2_0.message.order.Order;
import org.opentcs.commadapter.vehicle.vda5050.v2_0.message.state.State;
import org.opentcs.commadapter.vehicle.vda5050.v2_0.message.visualization.Visualization;

/**
 * A {@link JsonValidator} for VDA5050 1.1 messages.
 */
public class MessageValidator
    extends
      JsonValidator {

  /**
   * Creates a new instance.
   *
   * @throws IllegalArgumentException If there was a problem reading the VDA5050 schemas.
   */
  public MessageValidator()
      throws IllegalArgumentException {
    super(
        Map.of(
            Connection.class,
            new InputStreamReader(
                Connection.class.getResourceAsStream(Connection.JSON_SCHEMA_PATH)
            ),
            InstantActions.class,
            new InputStreamReader(
                InstantActions.class.getResourceAsStream(InstantActions.JSON_SCHEMA_PATH)
            ),
            Order.class,
            new InputStreamReader(
                Order.class.getResourceAsStream(Order.JSON_SCHEMA_PATH)
            ),
            State.class,
            new InputStreamReader(
                State.class.getResourceAsStream(State.JSON_SCHEMA_PATH)
            ),
            Visualization.class,
            new InputStreamReader(
                Visualization.class.getResourceAsStream(Visualization.JSON_SCHEMA_PATH)
            )
        )
    );
  }

}
