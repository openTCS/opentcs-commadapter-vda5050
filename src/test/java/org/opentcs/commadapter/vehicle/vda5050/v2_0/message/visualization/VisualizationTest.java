// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.commadapter.vehicle.vda5050.v2_0.message.visualization;

import java.time.Instant;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.opentcs.commadapter.vehicle.vda5050.ResourceLoader;
import org.opentcs.commadapter.vehicle.vda5050.common.JsonBinder;
import org.opentcs.commadapter.vehicle.vda5050.v2_0.MessageValidator;
import org.opentcs.commadapter.vehicle.vda5050.v2_0.message.common.AgvPosition;
import org.opentcs.commadapter.vehicle.vda5050.v2_0.message.common.Velocity;

/**
 * Unit tests for {@link Visualization}.
 */
public class VisualizationTest {

  private static final String RESOURCE_DIR
      = "/org/opentcs/commadapter/vehicle/vda5050/v2_0/message/visualization/";

  private MessageValidator messageValidator;
  private JsonBinder jsonBinder;

  @BeforeEach
  public void setUp() {
    messageValidator = new MessageValidator();
    jsonBinder = new JsonBinder();
  }

  @Test
  public void validateAgainstJsonSchemaMinimal() {
    messageValidator.validate(
        jsonBinder.toJson(createVisualizationMinimal()),
        Visualization.class
    );
  }

  @Test
  public void validateAgainstJsonSchemaFull() {
    messageValidator.validate(
        jsonBinder.toJson(createVisualizationFull()),
        Visualization.class
    );
  }

  @Test
  public void validateJsonWithNullForOptionalFieldsAgainstSchema() {
    messageValidator.validate(
        ResourceLoader.load(RESOURCE_DIR + "visualizationMessageWithNullForOptionalFields.json"),
        Visualization.class
    );
  }

  @Test
  public void deserializeJsonWithNullForOptionalFields() {
    Approvals.verify(
        jsonBinder.toJson(
            jsonBinder.fromJson(
                ResourceLoader.load(
                    RESOURCE_DIR + "visualizationMessageWithNullForOptionalFields.json"
                ),
                Visualization.class
            )
        )
    );
  }

  @Test
  public void validateJsonWithNullForOptionalRootFieldsAgainstSchema() {
    messageValidator.validate(
        ResourceLoader.load(
            RESOURCE_DIR + "visualizationMessageWithNullForOptionalRootFields.json"
        ),
        Visualization.class
    );
  }

  @Test
  public void deserializeJsonWithNullForOptionalRootFields() {
    Approvals.verify(
        jsonBinder.toJson(
            jsonBinder.fromJson(
                ResourceLoader.load(
                    RESOURCE_DIR + "visualizationMessageWithNullForOptionalRootFields.json"
                ),
                Visualization.class
            )
        )
    );
  }

  @Test
  public void jsonSampleMinimal() {
    Approvals.verify(jsonBinder.toJson(createVisualizationMinimal()));
  }

  @Test
  public void jsonSampleFull() {
    Approvals.verify(jsonBinder.toJson(createVisualizationFull()));
  }

  private Visualization createVisualizationMinimal() {
    return new Visualization(
        0L,
        Instant.EPOCH,
        "version",
        "manufacturer",
        "serial-number"
    );
  }

  private Visualization createVisualizationFull() {
    return new Visualization(
        0L,
        Instant.EPOCH,
        "version",
        "manufacturer",
        "serial-number"
    )
        .setAgvPosition(
            new AgvPosition(1.2, 3.4, 1.2, "some-map-id", true)
                .setDeviationRange(0.1)
                .setLocalizationScore(0.2)
                .setMapDescription("some-map-description")
        )
        .setVelocity(
            new Velocity()
                .setVx(0.1)
                .setVy(0.2)
                .setOmega(0.3)
        );
  }
}
