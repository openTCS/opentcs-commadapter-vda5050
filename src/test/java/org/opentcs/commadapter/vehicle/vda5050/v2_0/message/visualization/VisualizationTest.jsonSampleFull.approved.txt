{
  "headerId" : 0,
  "timestamp" : "1970-01-01T00:00:00Z",
  "version" : "version",
  "manufacturer" : "manufacturer",
  "serialNumber" : "serial-number",
  "agvPosition" : {
    "x" : 1.2,
    "y" : 3.4,
    "theta" : 1.2,
    "mapId" : "some-map-id",
    "positionInitialized" : true,
    "mapDescription" : "some-map-description",
    "localizationScore" : 0.2,
    "deviationRange" : 0.1
  },
  "velocity" : {
    "vx" : 0.1,
    "vy" : 0.2,
    "omega" : 0.3
  }
}